File: frontend/src/api/currency.ts
/**
 * Currency types and API functions.
 */

// API Configuration
const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL || "http://localhost:8000";

// Types
export interface CurrencyInfo {
  code: string;
  symbol: string;
  name: string;
}

export interface ExchangeRates {
  baseCurrency: string;
  rates: Record<string, number>;
  updatedAt: string | null;
}

export interface ExchangeRate {
  fromCurrency: string;
  toCurrency: string;
  rate: number;
}

export interface CurrencyPreferences {
  defaultCurrency: string;
  showConverted: boolean;
}

export interface ConversionResult {
  originalAmount: number;
  originalCurrency: string;
  convertedAmount: number;
  targetCurrency: string;
  rate: number;
  formattedOriginal: string;
  formattedConverted: string;
}

export interface BrokerCurrencies {
  broker: string;
  currencies: string[];
}

export interface AccountCurrency {
  accountId: number;
  accountName: string;
  broker: string;
  currency: string;
}

export interface FormattedAmount {
  original: {
    amount: number;
    currency: string;
    formatted: string;
  };
  converted: {
    amount: number;
    currency: string;
    formatted: string;
  } | null;
}

// Currency symbols mapping (for client-side formatting)
export const CURRENCY_SYMBOLS: Record<string, string> = {
  SEK: "kr",
  DKK: "kr",
  EUR: "€",
  USD: "$",
  GBP: "£",
  NOK: "kr",
  CHF: "CHF",
  JPY: "¥",
  AUD: "A$",
  CAD: "C$",
  NZD: "NZ$",
};

// Default values for fallback
const DEFAULT_CURRENCIES: CurrencyInfo[] = [
  { code: "SEK", symbol: "kr", name: "Swedish Krona" },
  { code: "DKK", symbol: "kr", name: "Danish Krone" },
  { code: "EUR", symbol: "€", name: "Euro" },
  { code: "USD", symbol: "$", name: "US Dollar" },
  { code: "GBP", symbol: "£", name: "British Pound" },
];

const DEFAULT_EXCHANGE_RATES: ExchangeRates = {
  baseCurrency: "SEK",
  rates: {
    SEK: 1.0,
    DKK: 1.52,
    EUR: 11.32,
    USD: 10.5,
    GBP: 13.2,
  },
  updatedAt: null,
};

const DEFAULT_PREFERENCES: CurrencyPreferences = {
  defaultCurrency: "SEK",
  showConverted: true,
};

// Fetch helper
async function fetchApi<T>(
  endpoint: string,
  options?: RequestInit,
): Promise<T> {
  const url = `${API_BASE_URL}${endpoint}`;

  const response = await fetch(url, {
    ...options,
    headers: {
      "Content-Type": "application/json",
      ...options?.headers,
    },
  });

  if (!response.ok) {
    throw new Error(`HTTP error ${response.status}`);
  }

  return response.json();
}

// API Functions

/**
 * Get list of all supported currencies.
 */
export async function getSupportedCurrencies(): Promise<CurrencyInfo[]> {
  try {
    return await fetchApi<CurrencyInfo[]>("/api/currency/supported");
  } catch {
    return DEFAULT_CURRENCIES;
  }
}

/**
 * Get list of currencies actually used in trading data.
 */
export async function getCurrenciesInUse(): Promise<string[]> {
  try {
    return await fetchApi<string[]>("/api/currency/in-use");
  } catch {
    return [];
  }
}

/**
 * Get all exchange rates relative to a base currency.
 */
export async function getExchangeRates(
  baseCurrency: string = "SEK",
): Promise<ExchangeRates> {
  try {
    return await fetchApi<ExchangeRates>(
      `/api/currency/rates?base=${baseCurrency}`,
    );
  } catch {
    return DEFAULT_EXCHANGE_RATES;
  }
}

/**
 * Get exchange rate between two specific currencies.
 */
export async function getExchangeRate(
  fromCurrency: string,
  toCurrency: string,
): Promise<ExchangeRate> {
  try {
    return await fetchApi<ExchangeRate>(
      `/api/currency/rates/${fromCurrency}/${toCurrency}`,
    );
  } catch {
    // Calculate from default rates
    const fromRate = DEFAULT_EXCHANGE_RATES.rates[fromCurrency] || 1;
    const toRate = DEFAULT_EXCHANGE_RATES.rates[toCurrency] || 1;
    return {
      fromCurrency,
      toCurrency,
      rate: fromRate / toRate,
    };
  }
}

/**
 * Update a single exchange rate.
 */
export async function updateExchangeRate(
  fromCurrency: string,
  toCurrency: string,
  rate: number,
): Promise<{ success: boolean; message: string }> {
  return fetchApi("/api/currency/rates", {
    method: "PUT",
    body: JSON.stringify({
      fromCurrency,
      toCurrency,
      rate,
    }),
  });
}

/**
 * Bulk update exchange rates.
 */
export async function bulkUpdateRates(
  rates: Record<string, number>,
  baseCurrency: string = "SEK",
): Promise<{ success: boolean; message: string }> {
  return fetchApi("/api/currency/rates/bulk", {
    method: "PUT",
    body: JSON.stringify({
      baseCurrency,
      rates,
    }),
  });
}

/**
 * Convert an amount between currencies.
 */
export async function convertCurrency(
  amount: number,
  fromCurrency: string,
  toCurrency: string,
): Promise<ConversionResult> {
  try {
    return await fetchApi("/api/currency/convert", {
      method: "POST",
      body: JSON.stringify({
        amount,
        fromCurrency,
        toCurrency,
      }),
    });
  } catch {
    // Calculate locally
    const rate = await getExchangeRate(fromCurrency, toCurrency);
    const converted = amount * rate.rate;
    return {
      originalAmount: amount,
      originalCurrency: fromCurrency,
      convertedAmount: converted,
      targetCurrency: toCurrency,
      rate: rate.rate,
      formattedOriginal: formatCurrency(amount, fromCurrency),
      formattedConverted: formatCurrency(converted, toCurrency),
    };
  }
}

/**
 * Convert an amount between currencies (GET version).
 */
export async function convertCurrencyGet(
  amount: number,
  fromCurrency: string,
  toCurrency: string,
): Promise<ConversionResult> {
  try {
    return await fetchApi<ConversionResult>(
      `/api/currency/convert?amount=${amount}&from=${fromCurrency}&to=${toCurrency}`,
    );
  } catch {
    return convertCurrency(amount, fromCurrency, toCurrency);
  }
}

/**
 * Get user currency preferences.
 */
export async function getCurrencyPreferences(): Promise<CurrencyPreferences> {
  try {
    return await fetchApi<CurrencyPreferences>("/api/currency/preferences");
  } catch {
    return DEFAULT_PREFERENCES;
  }
}

/**
 * Update user currency preferences.
 */
export async function updateCurrencyPreferences(
  prefs: CurrencyPreferences,
): Promise<{ success: boolean; preferences: CurrencyPreferences }> {
  try {
    return await fetchApi("/api/currency/preferences", {
      method: "PUT",
      body: JSON.stringify(prefs),
    });
  } catch {
    return { success: false, preferences: prefs };
  }
}

/**
 * Get the default display currency.
 */
export async function getDefaultCurrency(): Promise<{
  defaultCurrency: string;
}> {
  try {
    return await fetchApi<{ defaultCurrency: string }>(
      "/api/currency/preferences/default",
    );
  } catch {
    return { defaultCurrency: DEFAULT_PREFERENCES.defaultCurrency };
  }
}

/**
 * Set the default display currency.
 */
export async function setDefaultCurrency(
  currency: string,
): Promise<{ success: boolean; defaultCurrency: string }> {
  try {
    return await fetchApi(
      `/api/currency/preferences/default?currency=${currency}`,
      {
        method: "PUT",
      },
    );
  } catch {
    return { success: false, defaultCurrency: currency };
  }
}

/**
 * Get list of brokers with their associated currencies.
 */
export async function getBrokersWithCurrencies(): Promise<BrokerCurrencies[]> {
  try {
    return await fetchApi<BrokerCurrencies[]>("/api/currency/brokers");
  } catch {
    return [];
  }
}

/**
 * Get currencies associated with each account.
 */
export async function getAccountCurrencies(): Promise<AccountCurrency[]> {
  try {
    return await fetchApi<AccountCurrency[]>("/api/currency/accounts");
  } catch {
    return [];
  }
}

/**
 * Format a currency amount (server-side).
 */
export async function formatCurrencyServer(
  amount: number,
  currency: string,
  includeSymbol: boolean = true,
  decimalPlaces: number = 2,
): Promise<{ amount: number; currency: string; formatted: string }> {
  try {
    return await fetchApi(
      `/api/currency/format?amount=${amount}&currency=${currency}&includeSymbol=${includeSymbol}&decimalPlaces=${decimalPlaces}`,
    );
  } catch {
    return {
      amount,
      currency,
      formatted: formatCurrency(amount, currency, {
        includeSymbol,
        decimalPlaces,
      }),
    };
  }
}

/**
 * Format amount with optional conversion display (server-side).
 */
export async function formatWithConversion(
  amount: number,
  originalCurrency: string,
  targetCurrency?: string,
): Promise<FormattedAmount> {
  try {
    let url = `/api/currency/format-with-conversion?amount=${amount}&originalCurrency=${originalCurrency}`;
    if (targetCurrency) {
      url += `&targetCurrency=${targetCurrency}`;
    }
    return await fetchApi<FormattedAmount>(url);
  } catch {
    return {
      original: {
        amount,
        currency: originalCurrency,
        formatted: formatCurrency(amount, originalCurrency),
      },
      converted: null,
    };
  }
}

// Client-side formatting utilities

/**
 * Format currency amount on client side.
 */
export function formatCurrency(
  amount: number,
  currency: string,
  options?: {
    includeSymbol?: boolean;
    decimalPlaces?: number;
    locale?: string;
  },
): string {
  const {
    includeSymbol = true,
    decimalPlaces = 2,
    locale = "en-US",
  } = options || {};

  const formatted = new Intl.NumberFormat(locale, {
    minimumFractionDigits: decimalPlaces,
    maximumFractionDigits: decimalPlaces,
  }).format(amount);

  if (!includeSymbol) {
    return formatted;
  }

  const symbol = CURRENCY_SYMBOLS[currency] || currency;

  // Symbol placement varies by currency
  if (["USD", "GBP", "AUD", "CAD", "NZD"].includes(currency)) {
    return `${symbol}${formatted}`;
  }

  return `${formatted} ${symbol}`;
}

/**
 * Format currency with both original and converted amounts.
 */
export function formatWithConversionClient(
  amount: number,
  originalCurrency: string,
  targetCurrency: string,
  exchangeRate: number,
): FormattedAmount {
  const convertedAmount = amount * exchangeRate;

  return {
    original: {
      amount,
      currency: originalCurrency,
      formatted: formatCurrency(amount, originalCurrency),
    },
    converted:
      originalCurrency !== targetCurrency
        ? {
            amount: convertedAmount,
            currency: targetCurrency,
            formatted: formatCurrency(convertedAmount, targetCurrency),
          }
        : null,
  };
}

/**
 * Get currency symbol.
 */
export function getCurrencySymbol(currency: string): string {
  return CURRENCY_SYMBOLS[currency] || currency;
}

File: src/api/routers/currency.py
"""
Currency API router.

Provides endpoints for:
- Getting supported currencies
- Getting/updating exchange rates
- Managing user currency preferences
- Currency conversion
"""

from typing import Any, Dict, List, Optional

from fastapi import APIRouter, HTTPException, Query
from pydantic import BaseModel, Field

from api.services.currency import CurrencyService

router = APIRouter()


# Request/Response Models
class CurrencyInfo(BaseModel):
    """Currency information."""

    code: str
    symbol: str
    name: str


class ExchangeRateUpdate(BaseModel):
    """Exchange rate update request."""

    from_currency: str = Field(alias="fromCurrency")
    to_currency: str = Field(alias="toCurrency")
    rate: float

    class Config:
        populate_by_name = True


class BulkRatesUpdate(BaseModel):
    """Bulk exchange rates update request."""

    base_currency: str = Field(default="SEK", alias="baseCurrency")
    rates: Dict[str, float]

    class Config:
        populate_by_name = True


class CurrencyPreferences(BaseModel):
    """User currency preferences."""

    default_currency: Optional[str] = Field(default=None, alias="defaultCurrency")
    show_converted: bool = Field(default=True, alias="showConverted")

    class Config:
        populate_by_name = True


class ConversionRequest(BaseModel):
    """Currency conversion request."""

    amount: float
    from_currency: str = Field(alias="fromCurrency")
    to_currency: str = Field(alias="toCurrency")

    class Config:
        populate_by_name = True


class ConversionResult(BaseModel):
    """Currency conversion result."""

    original_amount: float = Field(alias="originalAmount")
    original_currency: str = Field(alias="originalCurrency")
    converted_amount: float = Field(alias="convertedAmount")
    target_currency: str = Field(alias="targetCurrency")
    rate: float
    formatted_original: str = Field(alias="formattedOriginal")
    formatted_converted: str = Field(alias="formattedConverted")

    class Config:
        populate_by_name = True


class BrokerCurrencies(BaseModel):
    """Broker with associated currencies."""

    broker: str
    currencies: List[str]


class AccountCurrency(BaseModel):
    """Account currency information."""

    account_id: int = Field(alias="accountId")
    account_name: str = Field(alias="accountName")
    broker: str
    currency: str

    class Config:
        populate_by_name = True


# Endpoints
@router.get("/supported", response_model=List[CurrencyInfo])
async def get_supported_currencies():
    """Get list of all supported currencies."""
    return CurrencyService.get_supported_currencies()


@router.get("/in-use", response_model=List[str])
async def get_currencies_in_use():
    """Get list of currencies actually used in trading data."""
    return CurrencyService.get_currencies_in_use()


@router.get("/rates")
async def get_exchange_rates(
    base: str = Query(default="SEK", description="Base currency for rates"),
):
    """Get all exchange rates relative to a base currency."""
    rates = CurrencyService.get_all_exchange_rates(base)
    return {
        "baseCurrency": base,
        "rates": rates,
        "updatedAt": None,  # Could add timestamp tracking
    }


@router.get("/rates/{from_currency}/{to_currency}")
async def get_exchange_rate(from_currency: str, to_currency: str):
    """Get exchange rate between two specific currencies."""
    rate = CurrencyService.get_exchange_rate(from_currency, to_currency)
    if rate is None:
        raise HTTPException(
            status_code=404,
            detail=f"Exchange rate not found for {from_currency} to {to_currency}",
        )
    return {
        "fromCurrency": from_currency,
        "toCurrency": to_currency,
        "rate": rate,
    }


@router.put("/rates")
async def update_exchange_rate(update: ExchangeRateUpdate):
    """Update a single exchange rate."""
    success = CurrencyService.update_exchange_rate(
        update.from_currency, update.to_currency, update.rate, source="api"
    )
    if not success:
        raise HTTPException(status_code=500, detail="Failed to update exchange rate")
    return {
        "success": True,
        "message": f"Updated rate: {update.from_currency} -> {update.to_currency} = {update.rate}",
    }


@router.put("/rates/bulk")
async def bulk_update_rates(update: BulkRatesUpdate):
    """Bulk update exchange rates relative to a base currency."""
    success = CurrencyService.bulk_update_rates(
        update.rates, update.base_currency, source="api"
    )
    if not success:
        raise HTTPException(status_code=500, detail="Failed to update exchange rates")
    return {
        "success": True,
        "message": f"Updated {len(update.rates)} exchange rates",
        "baseCurrency": update.base_currency,
    }


@router.post("/convert", response_model=ConversionResult)
async def convert_currency(request: ConversionRequest):
    """Convert an amount between currencies."""
    rate = CurrencyService.get_exchange_rate(request.from_currency, request.to_currency)
    if rate is None:
        raise HTTPException(
            status_code=400,
            detail=f"Cannot convert from {request.from_currency} to {request.to_currency}",
        )

    converted = request.amount * rate

    return ConversionResult(
        originalAmount=request.amount,
        originalCurrency=request.from_currency,
        convertedAmount=converted,
        targetCurrency=request.to_currency,
        rate=rate,
        formattedOriginal=CurrencyService.format_currency(
            request.amount, request.from_currency
        ),
        formattedConverted=CurrencyService.format_currency(
            converted, request.to_currency
        ),
    )


@router.get("/convert")
async def convert_currency_get(
    amount: float = Query(..., description="Amount to convert"),
    from_currency: str = Query(..., alias="from", description="Source currency"),
    to_currency: str = Query(..., alias="to", description="Target currency"),
):
    """Convert an amount between currencies (GET version)."""
    rate = CurrencyService.get_exchange_rate(from_currency, to_currency)
    if rate is None:
        raise HTTPException(
            status_code=400,
            detail=f"Cannot convert from {from_currency} to {to_currency}",
        )

    converted = amount * rate

    return {
        "originalAmount": amount,
        "originalCurrency": from_currency,
        "convertedAmount": converted,
        "targetCurrency": to_currency,
        "rate": rate,
        "formattedOriginal": CurrencyService.format_currency(amount, from_currency),
        "formattedConverted": CurrencyService.format_currency(converted, to_currency),
    }


@router.get("/preferences", response_model=CurrencyPreferences)
async def get_currency_preferences():
    """Get user currency preferences."""
    return CurrencyPreferences(
        defaultCurrency=CurrencyService.get_default_currency(),
        showConverted=CurrencyService.get_show_converted(),
    )


@router.put("/preferences")
async def update_currency_preferences(prefs: CurrencyPreferences):
    """Update user currency preferences."""
    success_currency = CurrencyService.set_default_currency(prefs.default_currency)
    success_converted = CurrencyService.set_show_converted(prefs.show_converted)

    if not success_currency:
        raise HTTPException(
            status_code=400,
            detail=f"Invalid currency: {prefs.default_currency}",
        )

    return {
        "success": True,
        "preferences": {
            "defaultCurrency": prefs.default_currency,
            "showConverted": prefs.show_converted,
        },
    }


@router.get("/preferences/default")
async def get_default_currency():
    """Get the default display currency."""
    return {
        "defaultCurrency": CurrencyService.get_default_currency(),
    }


@router.put("/preferences/default")
async def set_default_currency(currency: str = Query(..., description="Currency code")):
    """Set the default display currency."""
    success = CurrencyService.set_default_currency(currency)
    if not success:
        raise HTTPException(
            status_code=400,
            detail=f"Invalid or unsupported currency: {currency}",
        )
    return {
        "success": True,
        "defaultCurrency": currency,
    }


@router.get("/brokers", response_model=List[BrokerCurrencies])
async def get_brokers_with_currencies():
    """Get list of brokers with their associated currencies."""
    return CurrencyService.get_brokers_with_currencies()


@router.get("/accounts", response_model=List[AccountCurrency])
async def get_account_currencies():
    """Get currencies associated with each account."""
    accounts = CurrencyService.get_account_currencies()
    return [
        AccountCurrency(
            accountId=acc["account_id"],
            accountName=acc["account_name"],
            broker=acc["broker"],
            currency=acc["currency"],
        )
        for acc in accounts
    ]


@router.get("/format")
async def format_currency(
    amount: float = Query(..., description="Amount to format"),
    currency: str = Query(..., description="Currency code"),
    include_symbol: bool = Query(True, alias="includeSymbol"),
    decimal_places: int = Query(2, alias="decimalPlaces"),
):
    """Format a currency amount."""
    return {
        "amount": amount,
        "currency": currency,
        "formatted": CurrencyService.format_currency(
            amount, currency, include_symbol, decimal_places
        ),
    }


@router.get("/format-with-conversion")
async def format_with_conversion(
    amount: float = Query(..., description="Amount to format"),
    original_currency: str = Query(..., alias="originalCurrency"),
    target_currency: str = Query(
        ..., alias="targetCurrency", description="Target currency (required)"
    ),
):
    """Format amount with optional conversion display."""

    result = CurrencyService.format_with_conversion(
        amount, original_currency, target_currency
    )
    return result
